ecosystem/indexer-grpc/indexer-grpc-data-service-v2/src/live_data_service/in_memory_cache.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    connection_manager::ConnectionManager,
+    live_data_service::{data_manager::DataManager, fetch_manager::FetchManager},
+};
+use aptos_protos::transaction::v1::Transaction;
+use prost::Message;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+use tracing::trace;
+
+pub(super) struct InMemoryCache<'a> {
+    pub(super) data_manager: Arc<RwLock<DataManager>>,
+    pub(super) fetch_manager: Arc<FetchManager<'a>>,
+}
+
+impl<'a> InMemoryCache<'a> {
+    pub(super) fn new(
+        connection_manager: Arc<ConnectionManager>,
+        known_latest_version: u64,
+        num_slots: usize,
+        size_limit_bytes: usize,
+    ) -> Self {
+        let data_manager = Arc::new(RwLock::new(DataManager::new(
+            known_latest_version + 1,
+            num_slots,
+            size_limit_bytes,
+        )));
+        let fetch_manager = Arc::new(FetchManager::new(data_manager.clone(), connection_manager));
+        Self {
+            data_manager,
+            fetch_manager,
+        }
+    }
+
+    pub(super) async fn get_data(
+        &'a self,
+        starting_version: u64,
+        ending_version: u64,
+        max_num_transactions_per_batch: usize,
+        max_bytes_per_batch: usize,
+    ) -> Option<(Vec<Transaction>, usize)> {
+        while starting_version >= self.data_manager.read().await.end_version {
+            trace!("Reached head, wait...");
+            let num_transactions = self
+                .fetch_manager
+                .fetching_latest_data_task
+                .read()
+                .await
+                .as_ref()
+                .unwrap()
+                .clone()
+                .await;
+
+            trace!("Done waiting, got {num_transactions} transactions at head.");
+        }
+
+        loop {
+            let data_manager = self.data_manager.read().await;
+
+            trace!("Getting data from cache, requested_version: {starting_version}, oldest available version: {}.", data_manager.start_version);
+            if starting_version < data_manager.start_version {
+                return None;
+            }
+
+            let start_index = starting_version as usize % data_manager.num_slots;
+
+            if data_manager.data[start_index].is_none() {
```

Comment:
nit: roll this into a function; it's not clear to readers about how to access data_manager.

Reply:
Done